### PR TITLE
feat(ccp): ccp state persistence

### PR DESCRIPTION
**Pull request by monoid**
1. Store prover state to state directory on commitment.  Save empty state on no commitment.
2. On start, load the state and check proofs found.  Leave only valid proofs and use maximal valid proof index for further generation.

### CODE REVIEW

File: ccp/src/state_storage.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    pub(crate) cu_allocation: CUAllocation,
+}
+
+impl StateStorage {
+    pub(crate) fn new(state_dir: PathBuf) -> Self {
+        Self { state_dir }
+    }
+
+    pub(crate) async fn save_state(&self, state: Option<&CCPState>) -> tokio::io::Result<()> {
+        let data = serde_json::to_vec(&state).expect("TODO");
```
Comment (mikevoronov) [inline ccp/src/state_storage.rs:45]: let's write a suitable error message here

File: ccp/src/prover.rs [hunk 2]
```diff
@@ -127,6 +136,83 @@ impl CCProver {
...
+        let state_storage = StateStorage::new(config.dir_to_store_persistent_state);
+
+        let prev_state = state_storage.try_to_load_data().await?;
+
+        let start_proof_idx = proof_cleaner
+            .validate_proofs(prev_state.as_ref().map(|state| &state.epoch_params))
+            .await?;
+        log::info!("continuing from proof index {start_proof_idx}");
+
+        let cu_prover_config = CUProverConfig {
```
Comment (mikevoronov) [inline ccp/src/prover.rs:154]: `enable_msr` is missing here

File: ccp/src/prover.rs [hunk 3]
```diff
@@ -127,6 +136,83 @@ impl CCProver {
         // stop background thread
         self.utility_thread.stop().await.map_err(Into::into)
     }
+
+    pub async fn from_saved_state(
```
Comment (mikevoronov) [inline ccp/src/prover.rs:140]: Can we decouple the mutual part with `new` into a separate function and call it from these two?

File: ccp/src/prover.rs [hunk 4]
```diff
@@ -127,6 +136,83 @@ impl CCProver {
...
+            cu_prover_config,
+            status: CCStatus::Idle,
+            utility_thread,
+            proof_drainer: proof_cleaner,
+            state_storage,
+        };
+
+        if let Some(prev_state) = prev_state {
+            self_
+                .apply_commitment_parameters(prev_state.epoch_params, &prev_state.cu_allocation)
```
Comment (mikevoronov) [inline ccp/src/prover.rs:179]: Let's name it `apply_cc_parameters`, it'll be more corresponding with other parts of the code base.

File: ccp/src/state_storage.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        let path = self.state_dir.join(STATE_FILE);
+
+        log::info!("Saving state to {:?}", path);
+
+        tokio::task::spawn_blocking(move || save_reliably(&path, &data))
+            .await
+            .unwrap()
```
Comment (mikevoronov) [inline ccp/src/state_storage.rs:52]:
```suggestion
        tokio::task::spawn_blocking(move || save_reliably(&path, &data)).await?
```

File: ccp/src/utility_thread/proof_storage.rs [hunk 6]
```diff
@@ -35,6 +40,44 @@ impl ProofStorage {
     pub async fn store_new_proof(&self, proof: CCProof) -> tokio::io::Result<()> {
         let proof_as_string = serde_json::to_string(&proof).expect(EXPECT_DEFAULT_SERIALIZER);
         let proof_path = self.proof_directory.join(proof.id.idx.to_string());
-        tokio::fs::write(proof_path, proof_as_string).await
+        tokio::task::spawn_blocking(move || save_reliably(&proof_path, proof_as_string))
+            .await
+            // unwrap spawn_blocking outer result: if something is wrong here,
+            // the system is screwed.
+            .unwrap()
```
Comment (mikevoronov) [inline ccp/src/utility_thread/proof_storage.rs:47]:
```suggestion
        tokio::task::spawn_blocking(move || save_reliably(&proof_path, proof_as_string))
            .await?
```

File: ccp/src/utility_thread/proof_storage.rs [hunk 7]
```diff
@@ -35,6 +40,44 @@ impl ProofStorage {
...
+
+// This is a non-sync function to avoid possible tokio peculiarities.
+pub(crate) fn save_reliably(path: &Path, contents: impl AsRef<[u8]>) -> std::io::Result<()> {
+    // We might use random name here.  But if the dir is lock'd, it is not necessary.
+    let extension = draft_extension(path);
+    let mut draft_path = path.to_owned();
+    draft_path.set_extension(extension);
+
+    let base_dir = path.parent().map(Cow::Borrowed).unwrap_or_default();
+    // Yep; we open a directory as a file.  It's ok.
```
Comment (mikevoronov) [inline ccp/src/utility_thread/proof_storage.rs:59]: let's avoid such comments

File: ccp/src/utility_thread/proof_storage.rs [hunk 8]
```diff
@@ -35,6 +40,44 @@ impl ProofStorage {
...
+    // Now we have the contents saved reliably on disk as a draft file.
+    // If we rename, destination file will be changed atomically.
+    std::fs::rename(draft_path, path)?;
+    // Make sure rename is saved.
+    base_dir_file.sync_all()?;
+
+    Ok(())
+}
+
+fn draft_extension(path: &Path) -> OsString {
```
Comment (mikevoronov) [inline ccp/src/utility_thread/proof_storage.rs:78]: let's name it with a verb like `add_draft_extension`

File: main/src/main.rs [hunk 9]
```diff
@@ -148,7 +148,11 @@ fn build_prover(prover_args: ProverArgs) -> CCProver {
         enable_msr: prover_args.enable_msr,
     };
 
-    CCProver::new(prover_args.utility_core_id.into(), config)
+    CCProver::from_saved_state(prover_args.utility_core_id.into(), config)
+        .await
+        // e doesn't implemenet Sync, and cannot be converted to anyhow::Error or eyre::Error.
```
Comment (mikevoronov) [inline main/src/main.rs:153]:
```suggestion
        // e doesn't implement Sync, and cannot be converted to anyhow::Error or eyre::Error.
```

File: ccp/src/prover.rs (hunk 3 above)
Comment (monoid) [reply ccp/src/prover.rs:140]: It was my initial idea, but I failed to set idx counter inside the `NewProofHandler` that is very hermetic (and owned by a closure).

File: ccp/src/state_storage.rs (hunk 5 above)
Comment (monoid) [reply ccp/src/state_storage.rs]: Implemented.

File: ccp/src/utility_thread/proof_storage.rs (hunk 6 above)
Comment (monoid) [reply ccp/src/utility_thread/proof_storage.rs]: Implemented.

File: main/src/main.rs (hunk 9 above)
Comment (monoid) [reply main/src/main.rs]: Implemented.

**Review by mikevoronov — APPROVED** [review]
utACK

**Pull request merged**